### PR TITLE
Updates socket.io network extension to correctly detect and manage connection to socket.io server versions 0.9.x - 1.x

### PR DESCRIPTION
**Pull request by himynameschris**
Can be tested using: https://github.com/hannon235/socket.io-testserver.git
OR
Test version of 0.9.x socket.io server running at ws://dev.channon.us:3009
Test version of 1.x socket.io server running at ws://dev.channon.us:3010

### CODE REVIEW

**Comment by eltonjothi** [comment]
Just tested this with jsb and it works
thanks a lot :+1:

**Comment by zilongshanren** [comment]
@samuele3hu  please take a look at this  [ci rebuild]

**Comment by danutchereches** [comment]
I think there's a problem with the "switch (_version)" (line 764).
"case 5" and "case 6" should probably be part of "switch (control)". (line 863)
It looks like misplaced braces.

**Comment by himynameschris** [comment]
@loadrunner you are correct about the switch statements, ive fixed this. long term goal should be to replace these with individual functions and avoid the nested switches

@andyque @samuele3hu I've also pulled in the most recent commits from the v3 branch and updated the PR branch with those

**Comment by samuele3hu** [comment]
@hannon235, i will test it in v3.8,thanks.

File: cocos/network/SocketIO.cpp [hunk 1]
```diff
@@ -36,12 +36,324 @@
 #include "WebSocket.h"
 #include "HttpClient.h"
 
+#include "json/rapidjson.h"
```
Comment (zilongshanren) [inline cocos/network/SocketIO.cpp:39]: Why include the rapidjson files twice? Typo here?

File: cocos/network/SocketIO.cpp [hunk 2]
```diff
@@ -36,12 +36,324 @@
...
 namespace network {
 
 //class declarations
 
+class SocketIOPacketV10x;
+
+class SocketIOPacket
+{
+public:
+    typedef enum
```
Comment (zilongshanren) [inline cocos/network/SocketIO.cpp:60]: in c++11, strong typed enum is preferred.

File: cocos/network/SocketIO.cpp [hunk 3]
```diff
@@ -36,12 +36,324 @@
...
+
+        outS = s.GetString();
+
+        CCLOGINFO("create args object: %s:", outS.c_str());
+    }
+
+    return outS;
+}
+
+SocketIOPacketV10x::SocketIOPacketV10x()
```
Comment (zilongshanren) [inline cocos/network/SocketIO.cpp:245]: please use class initialize list to init all of the class's member vars.

File: cocos/network/SocketIO.cpp [hunk 4]
```diff
@@ -36,12 +36,324 @@
...
+public:
+    SocketIOPacketV10x();
+    virtual ~SocketIOPacketV10x();
+    int typeAsNumber();
+    std::string stringify();
+private:
+    std::vector<std::string> _typesMessage;
+};
+
+SocketIOPacket::SocketIOPacket()
```
Comment (zilongshanren) [inline cocos/network/SocketIO.cpp:109]: please use class initialize list to init class member vars.

**Comment by zilongshanren** [comment]
@hannon235  Thanks for your contribution. I will do some refactoring to your code to meet our c++ coding style.

**Comment by zilongshanren** [comment]
@hannon235  You could just ignore these comments in this PR。

**Comment by zilongshanren** [comment]
@hannon235  My PR is here: https://github.com/cocos2d/cocos2d-x/pull/12740

**Comment by Longpc** [comment]
please add some test case! By this code. I can connect to NodeJs server 1.x but client->emit(...) can't send enything to server

**Comment by zilongshanren** [comment]
@hannon235  Thoughts?
@Longpc  I'm sorry for the issue, but I have tested all the tests bundled with cocos2d-x and it works well.

**Comment by himynameschris** [comment]
@Longpc test cases exist in this pull request, check the file tests/cpp-tests/Classes/ExtensionsTest/NetworkTest/SocketIOTest.cpp

If you are having issues, can you provide the client and server code you are using?

**Comment by Longpc** [comment]
OK I got it. @hannon235 @zilongshanren Thanks for your reply

**Comment by Longpc** [comment]
@hannon235 I have pull newest version in this repo. Change the url in your test to my server.
`_sioClient = SocketIO::connect("ws://192.168.0.198:8000", *this);`
It can connect success to server but the client->emit() still same issue.
This is my server code

    var server = require('http').Server();

    var socketIO = require('socket.io');

    var io = socketIO.listen(server);

    io.sockets.on("connection", function(socket) {
	console.log("connect from client with ID" + socket.id);
	
	socket.emit("testevent", "Wellcome to NodeJs sv");
	
	socket.on('echotest', function(data) {
		console.log("event from client with data" + data);
		var data = JSON.parse(data);
		socket.emit("echotest", data);
	});
	
	socket.on("disconnect", function(data) {
		console.log(socket.id + " disconnected");
		console.log(data);
	});
    });

    server.listen('8000');
    console.log("server listen on port 8000");
and here is server log:
sv
an client log:
cl
By look at that. connect() function has no issue and client->on() too. But client->emit() has something wrong.

**Comment by Longpc** [comment]
May be I have misstake in git command.
because in last PR of this issue. Here is `SocketIOPacketV10x::SocketIOPacketV10x()` function  in  `SocketIO.cpp` file. 
pull
but in branch v3 of `cocos2d/coco2d-x` repo this pie of code become
rep
So my local repo same with this and the emit() function cannot execute.
I have change this code same with PR and it work perfectly.
Thanks you again and again. @hannon235 @zilongshanren

**Comment by roed** [comment]
still, the library doesn't support callback methods, which I added in the first commit (the pull request that was compatible with 1.x only: https://github.com/cocos2d/cocos2d-x/pull/10033).

socket.io has a packet for that, the one that only has a CCLOGINFO("Message Ack"); in the current code. Are you planning on adding that?

File: cocos/network/SocketIO.cpp [hunk 5]
```diff
@@ -116,10 +430,10 @@ SIOClientImpl::~SIOClientImpl()
 
 void SIOClientImpl::handshake()
 {
-    log("SIOClientImpl::handshake() called");
+    CCLOGINFO("SIOClientImpl::handshake() called");
 
     std::stringstream pre;
-    pre << "http://" << _uri << "/socket.io/1";
+    pre << "http://" << _uri << "/socket.io/1/?EIO=2&transport=polling&b64=true";
```
Comment (pandamicro) [inline cocos/network/SocketIO.cpp:436]: Hi, @himynameschris 

Our team report a bug on Android shows that this URL is not reachable in the server. Indeed `/socket.io/1/` is not a regular path on any server. I couldn't understand why handshake need such a strange path, so I looked back and found the initial implementation was contributed by you.

I want to ask, why this specific path for handshake ?

Comment (pandamicro) [inline cocos/network/SocketIO.cpp:436]: I just re verified, there was no problem, even if the url doesn't exist, the handshake can pass. What we met was another problem, so please ignore it

**Pull request merged**